File: JavaHashGen/XHash.java
package JavaHashGen;

/** 
 * Password Hashing With PBKDF2 (http://crackstation.net/hashing-security.htm).
 * Copyright (c) 2013, Taylor Hornby
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, 
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation 
 * and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 * POSSIBILITY OF SUCH DAMAGE.
 */

import java.security.SecureRandom;

import javax.crypto.spec.DESKeySpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.Cipher;
import javax.crypto.SecretKeyFactory;

import com.sun.org.apache.xerces.internal.impl.dv.util.Base64;

import java.math.BigInteger;
import java.security.NoSuchAlgorithmException;
import java.security.spec.AlgorithmParameterSpec;
import java.security.spec.InvalidKeySpecException;

/**
 * PBKDF2 salted password hashing. Author: havoc AT defuse.ca www:
 * http://crackstation.net/hashing-security.htm
 */
public class XHash {
	private static final String PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA1";

	// The following constants may be changed without breaking existing hashes.
	private static final int SALT_BYTE_SIZE = 24;
	private static final int HASH_BYTE_SIZE = 24;
	private static final int PBKDF2_ITERATIONS = 5000;

	private static final int SALT_LENGTH = 32;
	private static final int PBKDF2_LENGTH = 32;

	/**
	 * Returns a salted PBKDF2 hash of the password.
	 * 
	 * @param password
	 *            the password to hash
	 * @return a salted PBKDF2 hash of the password
	 */
	public static String create(String input) 
		throws NoSuchAlgorithmException, InvalidKeySpecException 
	{
		if(input==null||input.trim().equals(""))
			return null;
		return create((input).toCharArray());
	}

	/**
	 * Returns a salted PBKDF2 hash of the password.
	 * 
	 * @param password
	 *            the password to hash
	 * @return a salted PBKDF2 hash of the password
	 */
	private static String create(char[] password) 
		throws NoSuchAlgorithmException, InvalidKeySpecException 
	{
		// Generate a random salt
		SecureRandom random = new SecureRandom();
		byte[] salt = new byte[SALT_BYTE_SIZE];
		random.nextBytes(salt);

		// Hash the password
		byte[] hash = pbkdf2(password, salt, PBKDF2_ITERATIONS, HASH_BYTE_SIZE);
		return Base64.encode(salt) + Base64.encode(hash);
	}
	
	/**
	 * Validates a password given a hash of the correct one.
	 * 
	 * @param input
	 *            The password to check
	 * @param correctHash
	 *            A hash of the correct password
	 * @return a salted PBKDF2 hash of the password
	 */
	public static boolean validate(String input, String correctHash)
		throws NoSuchAlgorithmException, InvalidKeySpecException 
	{
		if(input==null||input.trim().equals(""))
			return false;
		if(correctHash==null||correctHash.trim().equals(""))
			return false;
		return validatePassword(input.toCharArray(), correctHash);
	}

	/**
     * Validates a password using a hash.
     *
     * @param   password        the password to check
     * @param   correctHash     the hash of the valid password
     * @return                  true if the password is correct, false if not
     */
    private static boolean validatePassword(char[] password, String correctHash)
        throws NoSuchAlgorithmException, InvalidKeySpecException
    {
        // Decode the hash into its parameters
        byte[] salt = Base64.decode(correctHash.substring(0,SALT_LENGTH));
        byte[] hash = Base64.decode(correctHash.substring(SALT_LENGTH));
        // Compute the hash of the provided password, using the same salt, 
        // iteration count, and hash length
        byte[] testHash = pbkdf2(password, salt, PBKDF2_ITERATIONS, HASH_BYTE_SIZE);
        // Compare the hashes in constant time. The password is correct if
        // both hashes match.
        return slowEquals(hash, testHash);
    }

	/**
	 * Computes the PBKDF2 hash of a password.
	 * 
	 * @param password
	 *            the password to hash.
	 * @param salt
	 *            the salt
	 * @param iterations
	 *            the iteration count (slowness factor)
	 * @param bytes
	 *            the length of the hash to compute in bytes
	 * @return the PBDKF2 hash of the password
	 */
	private static byte[] pbkdf2(char[] password, byte[] salt, int iterations, int bytes) 
		throws NoSuchAlgorithmException, InvalidKeySpecException 
	{
		PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, bytes * 8);
		SecretKeyFactory skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
		return skf.generateSecret(spec).getEncoded();
	}

	/**
     * Compares two byte arrays in length-constant time. This comparison method
     * is used so that password hashes cannot be extracted from an on-line 
     * system using a timing attack and then attacked off-line.
     * 
     * @param   a       the first byte array
     * @param   b       the second byte array 
     * @return          true if both byte arrays are the same, false if not
     */
    private static boolean slowEquals(byte[] a, byte[] b)
    {
        int diff = a.length ^ b.length;
        for(int i = 0; i < a.length && i < b.length; i++)
            diff |= a[i] ^ b[i];
        return diff == 0;
    }

	///**
	// * Tests the basic functionality of the PasswordHash class
	// * 
	// * @param args
	// *            ignored
	// */
	//public static void main(String[] args) {
	//	 try {
	//		 String tick="123456";
	//		 System.out.println(create(tick));	
	//	 } catch (Exception ex) {
	//		 System.out.println("ERROR: " + ex);
	//	 }
	// }
}

File: CSharpHashGen/XHashChecker/XHash.cs
using System;
using System.Security.Cryptography;

namespace XHashChecker
{
    /// <summary>
    /// Salted password hashing with PBKDF2-SHA1.
    /// Author: havoc AT defuse.ca
    /// www: http://crackstation.net/hashing-security.htm
    /// Compatibility: .NET 3.0 and later.
    /// </summary>
    public static class XHash
    {
        // The following constants may be changed without breaking existing hashes.
        private const int SALT_BYTE_SIZE = 24;
        private const int HASH_BYTE_SIZE = 24;
        private const int PBKDF2_ITERATIONS = 5000;
        private const int PBKDF2_AUTHORIZE_ITERATIONS = 50;

        private const int SALT_LENGTH = 32;
        private const int PBKDF2_LENGTH = 32;

        /// <summary>
        /// Creates a salted PBKDF2 hash of the password.
        /// </summary>
        /// <param name="input">The password to hash.</param>
        /// <returns>The hash of the password.</returns>
        public static string Create(string input)
        {
            if (string.IsNullOrWhiteSpace(input))
            {
                return null;
            }
            // Generate a random salt
            RNGCryptoServiceProvider csprng = new RNGCryptoServiceProvider();
            byte[] salt = new byte[SALT_BYTE_SIZE];
            csprng.GetBytes(salt);

            // Hash the password and encode the parameters
            byte[] hash = Pbkdf2(input, salt, PBKDF2_ITERATIONS, HASH_BYTE_SIZE);
            return Convert.ToBase64String(salt) + Convert.ToBase64String(hash);
        }

        /// <summary>
        /// Creates a salted PBKDF2 hash of the password.
        /// </summary>
        /// <param name="tick"></param>
        /// <param name="publicKey"></param>
        /// <returns>The hash of the password.</returns>
        public static string CreateAuthorize(string tick, string publicKey)
        {
            if (string.IsNullOrWhiteSpace(tick) || string.IsNullOrWhiteSpace(publicKey))
            {
                return null;
            }
            // Generate a random salt
            RNGCryptoServiceProvider csprng = new RNGCryptoServiceProvider();
            byte[] salt = new byte[SALT_BYTE_SIZE];
            csprng.GetBytes(salt);

            var input = tick + publicKey;
            // Hash the password and encode the parameters
            byte[] hash = Pbkdf2(input, salt, PBKDF2_AUTHORIZE_ITERATIONS, HASH_BYTE_SIZE);
            return Convert.ToBase64String(salt) + Convert.ToBase64String(hash);
        }

        /// <summary>
        /// Validates a password given a hash of the correct one.
        /// </summary>
        /// <param name="tick"></param>
        /// <param name="publicKey">The password to check.</param>
        /// <param name="correctHash">A hash of the correct password.</param>
        /// <returns>True if the password is correct. False otherwise.</returns>
        public static bool ValidateAuthorize(string tick, string publicKey, string correctHash)
        {
            if (string.IsNullOrWhiteSpace(tick) ||
                string.IsNullOrWhiteSpace(publicKey) ||
                string.IsNullOrWhiteSpace(correctHash))
            {
                return false;
            }
            var input = tick + publicKey;
            return Validate(input, correctHash, PBKDF2_AUTHORIZE_ITERATIONS);
        }
        /// <summary>
        /// Validates a password given a hash of the correct one.
        /// </summary>
        /// <param name="input">The password to check.</param>
        /// <param name="correctHash">A hash of the correct password.</param>
        /// <returns>True if the password is correct. False otherwise.</returns>
        public static bool Validate(string input, string correctHash)
        {
            if (string.IsNullOrWhiteSpace(input) ||
                string.IsNullOrWhiteSpace(correctHash))
            {
                return false;
            }
            return Validate(input, correctHash, PBKDF2_ITERATIONS);
        }

        /// <summary>
        /// Validates a password given a hash of the correct one.
        /// </summary>
        /// <param name="input">The password to check.</param>
        /// <param name="correctHash">A hash of the correct password.</param>
        /// <param name="iterations"></param>
        /// <returns>True if the password is correct. False otherwise.</returns>
        private static bool Validate(string input, string correctHash, int iterations)
        {
            if (correctHash.Length != SALT_LENGTH + PBKDF2_LENGTH)
            {
                return false;
            }
            // Extract the parameters from the hash
            var index = 0;
            byte[] salt = Convert.FromBase64String(correctHash.Substring(index, SALT_LENGTH));
            index += SALT_LENGTH;
            byte[] hash = Convert.FromBase64String(correctHash.Substring(index, PBKDF2_LENGTH));

            byte[] testHash = Pbkdf2(input, salt, iterations, hash.Length);
            return SlowEquals(hash, testHash);
        }

        /// <summary>
        /// Compares two byte arrays in length-constant time. This comparison
        /// method is used so that password hashes cannot be extracted from
        /// on-line systems using a timing attack and then attacked off-line.
        /// </summary>
        /// <param name="a">The first byte array.</param>
        /// <param name="b">The second byte array.</param>
        /// <returns>True if both byte arrays are equal. False otherwise.</returns>
        private static bool SlowEquals(byte[] a, byte[] b)
        {
            uint diff = (uint)a.Length ^ (uint)b.Length;
            for (int i = 0; i < a.Length && i < b.Length; i++)
                diff |= (uint)(a[i] ^ b[i]);
            return diff == 0;
        }

        /// <summary>
        /// Computes the PBKDF2-SHA1 hash of a password.
        /// </summary>
        /// <param name="password">The password to hash.</param>
        /// <param name="salt">The salt.</param>
        /// <param name="iterations">The PBKDF2 iteration count.</param>
        /// <param name="outputBytes">The length of the hash to generate, in bytes.</param>
        /// <returns>A hash of the password.</returns>
        private static byte[] Pbkdf2(string password, byte[] salt, int iterations, int outputBytes)
        {
            Rfc2898DeriveBytes pbkdf2 = new Rfc2898DeriveBytes(password, salt);
            pbkdf2.IterationCount = iterations;
            return pbkdf2.GetBytes(outputBytes);
        }
    }
}